### ISSUE TITLE
Create rego code-documentation support to support generating documentation about policies

### ISSUE THREAD

**Issue by petemounce** [open-policy-agent/opa#2180]
https://openpolicyagent.slack.com/archives/C1H19LW4F/p1583922893200700

I have a policy like

```rego
# Every kind should have the app label
deny[msg] {
	not common.has_field(input.metadata.labels, "app")
	msg = sprintf("%s[%s] missing 'app' label. Has: %s", [input.kind, name, input.metadata.labels])
}
```

and I'd like to be able to extract documentation from the rego code (or, alternatively, associated closely _with_ it, say in a next-door file) and generate documentation for publishing. That to me means extracting the comments into markdown, then transforming that into static site generator of choice.

I'd want to be able to:
* give a policy a stable unique ID, so I can deep-link/index into my docs site easily (shellcheck does this very well, example: https://github.com/koalaman/shellcheck/wiki/SC1000).
* have support for defining the why of a policy in prose.
* have support for defining what "good" might look like, with ability to have a code snippet block.
* ditto, for defining examples of what "bad" might look like (although I suppose that's the policy rego itself).
* ask within a code review for documentation to accompany changes to regos (regi?).

I think this would also enable others to share policies more readily.

**Comment by itaysk** [comment]
Hi, I also have interest in this. We are starting to add a header to our policies with metadata about the policy. For us this is about documentation but also about integration into other tools (the system that uses those rego files). 
We ended up defining something similar to a yaml front matter. We added it as a header to every file in a comment block. Formalizing this would be nice.
If this ever gets implemented, please consider allowing custom fields.

**Comment by itaysk** [comment]
as for good/bad examples - I think that a more appropriate place for those are in the policy tests. I personally liked [how it's done in golang](https://blog.golang.org/examples). Perhaps we could extent the test facility with "examples" semantics as well.

**Comment by tsandall** [comment]
We chatted about this on the community call yesterday. @johnharris85 did a neat demo of a tool they're building around OPA that implements some of these ideas.

One topic of discussion was how metadata should be defined. @jaspervdj-luminal pointed out that you can define the metadata directly in Rego. For example:

**deny.rego** (original rule)

```
package kubernetes.metadata.requireapplabel

deny[msg] {
	not common.has_field(input.metadata.labels, "app")
	msg = sprintf("%s[%s] missing 'app' label. Has: %s", [input.kind, name, input.metadata.labels])
}
```

**metadata.rego** (same package, different file, contains metadata about the policy)

```
package kubernetes.metadata.labelrequirements


title := "Label Requirements"

description := `The Label Requirements policy ensures
that all resources include an 'app' label for tracking
purposes`

tags := {"code": "SC1000"}

severity := data.policy.constants.CRITICAL
```

**examples.rego** (again, another file that contains example data)

```
package kubernetes.metadata.labelrequirements

example_good := {"metadata": {"labels": {"app": "nginx"}}}

example_bad := {"metadata": {"labels": {"owner": "bob"}}}
```

The package provides a namespace for the policy. Here we've got a 3-level namespace (domain=kubernetes, subdomain=metadata, policy=requireapplabel). Metadata can be defined in the same file or another. You can use raw strings for multi-line descriptions. Markdown could be embedded in those strings. Examples would require some convention (similar to tests like in Go).

A small library could be written on top of the github.com/open-policy-agent/opa/rego package to extract the metadata, examples, etc.

**Comment by johnharris85** [comment]
I've been playing with migrating from the approached demo'd (`#`-style doc strings) to this rego-native approach, although the multi-line string support is a bit limiting in terms of what Markdown is possible. For example code highlighting / fences are out because ` is used as the raw-string delimeter so something like the following errors:

```
description := `blah blah
    ```
    code fence
    ```
    blah blah

    blah blah.`
```

Could put the whole thing inside a single string with the appropriate newline characters and spaces but that ends up looking pretty ugly. Alternatively I could come up with some custom Markdown hack where ` is replaced with some other character and then flipped back later when I need to render it, but that feels a bit hacky and un-intuitive to the user. Anyone have any thoughts / ideas here?

**Comment by tsandall** [comment]
@johnharris85 I definitely agree that custom Markdown seems like a bad idea. Using a regular string w/ newline characters will be horrendous. A relatively simple solution (to implement) would be a new syntax for multi-line strings. Some thought would need to go into syntax (need to be careful not to overlap w/ anything else) and whether it's just an alternative to backtick (i.e., it generates a raw string) or whether there's anything else that should be done.

**Comment by itaysk** [comment]
In the last community call we discussed this issue and I drafted a proposal for your review. Appreciate your feedback: https://hackmd.io/@ZtQnh19kS26YiNlJLqKJnw/H1gAv5nBw

**Comment by patrick-east** [comment]
Thanks for putting that together @itaysk 👍 

One thing I really like about it is that it seems like that could pretty easily be done without any changes to OPA, as a purely convention-based thing + tooling that supports it, as a way to get things started.

**Comment by itaysk** [comment]
Thanks @patrick-east that was my aim. I'm hoping to discuss this in the next community call

**Comment by itaysk** [comment]
I've maintained the proposal here so far: https://hackmd.io/k6k7TVkFQ2m-ElQ2OT7nBQ and in the last community call we decided to bring it back to GitHub. The document is still available and includes further background, context, references, use cases, detailed considerations for every decision made. I'm pasting here just the spec section from the document as a snapshot to pick up the discussion from:

<details>
  <summary>Regodoc spec</summary>

### Policy metadata

1. A policy can declare metadoc information by declaring a **rego rule**.
2. The rule must be in the **same package** as the policy it's describing.
2. The rule name must be **`__rego__metadoc__`**.
4. The rule must be a **JSON literal** that may contain any of the supported fields described below.
5. It's recommended but not required that the metadoc will be defined at the top of the file.

Supported fields:

field | description | type (validation) | example
--- | --- | --- | ---
id | Unique identifier for the policy | string | XYZ
title | Descriptive name for the policy | string | Pod has privileged container
description | Full description of the policy goal and business logic. Ideally not covering it's usage (which is described separately) | string (markdown) | This policy checks for containers in the Pod that have the `securityContext.privileged` field set to `true`. Privileged containers can allow almost completely unrestricted host access.
custom | Additional information about the policy. For example: remediation advice, references, etc. | object | <table><tr><th>key</th><th>value</th></tr><tr><td>severity</td><td>High</td></tr><tr><td>links</td><td><ul><li>[Kubernetes Docs: Privileged mode for pod containers](https://kubernetes.io/docs/concepts/workloads/pods/pod/#privileged-mode-for-pod-containers)</li></ul></td></tr></table>
entrypoints | Description of existing rules that should be queried. Further information can be described in the entrypoints ruledoc | array of string (path to rule) | app.xyz

Example:

```rego
package app.xyz

__rego__metadoc__ := {
  "id": "XYZ",
  "title": "Pod has privileged container",
  "description": "This policy checks for containers in the Pod that have the `securityContext.privileged` field set to `true`. Privileged containers can allow almost completely unrestricted host access.",
  "custom": {
    "severity": "High",
    "links": [
      {
        "title": "Kubernetes Docs: Privileged mode for pod containers",
        "href": "https://kubernetes.io/docs/concepts/workloads/pods/pod/#privileged-mode-for-pod-containers)"
      }  
    ]
  },
  "entrypoints" : [
      "app.xyz.deny"
  ]
}

deny[ret] {
  ...
}
```

### Rule documentation

1. A policy can declare rule level code documentation by declaring an **incremental object rule**. By using an increments rule, each rule's documentation can be adjacent to it's defintion.
2. The object keys represent existing rules, and the value is that rule's documentation.
3. The rule must be in the **same package** as the policy it's describing.
4. The rule name must be **`__rego__ruledoc__`**.
5. The rule must be a **JSON literal** that may contain any of the supported fields described below.
6. It's recommended but not requried that the ruledoc will be defined right before the rule it's documenting.

Supported fields:

field | description | type (validation) | example
--- | --- | --- | ---
description | Full description of the rule's goal and business logic | string (markdown) | containers returns all the containers in a Kubernetes resource (Pod, Deployment, etc.), if any. This includes init containers and regular containers.
input | Description of this rule's expecations of the input document, or incase of a function, it's arguments as well | string | the input docuemnt should be a valid Kubernetes resouce, or a valid Kubernetes Admission Review resource, of Pod, or another resource that defines a Pod, such as Deployment, ReplicaSet, StatefulSet, etc.
input_schema | URL of JSON schema describing the input | http://path/to/jsonschema
output | Description of this rule's result | string | set of container objects
examples | Full paths to tests that demonstrate basic usage of the rule | array of string (path to rule) | <ul><li>app.xyz.test_privileged_denied</li><li>app.xyz.test_unprivileged_allowed</li></ul>

Example:

```rego
package app.xyz

__rego__ruledoc__["deny"] = {
    "description": "containers returns all the containers in a Kubernetes resource (Pod, Deployment, etc.), if any. This includes init containers and regular containers",
    "input": "the input docuemnt should be a valid Kubernetes resouce, or a valid Kubernetes Admission Review resource, of Pod, or another resource that defines a Pod, such as Deployment, ReplicaSet, StatefulSet, etc.",
    "input_schema": "https://raw.githubusercontent.com/instrumenta/kubernetes-json-schema/master/v1.18.1-standalone/pod.json",
    "output": "set of container objects",
    "examples": [
        "app.xyz.test_privileged_denied",
        "app.xyz.test_unprivileged_allowed"
    ]
}

deny[res] {
  ...
}
```

### Language syntax

As a second phase, it would be nice to extend the basic spec described here with "syntactic sugar" that will make metadoc authoring more streamlined, and documented policies more readable. This section describes a syntax that can be transformed into the primitive metadoc/ruledoc described above. The proposed syntax takes the following form:

Syntax spec:
1. A policy author can add a documentation block instead of defining documentation using a rule. the result is the same.
2. Each block defines either a `__rego__metadoc__` object or a `__rego__ruledoc__` object which is determined by the position of the block: 
    1. If the block is located right before the `package` declaration, it is defining policy metadata.
    2. If the block is located right before the rule head, it is defining that rule's documentation.
4. A block is a collection of lines. Each line must start with the character `@`.
6. Each line sets one field in the generated object, by using a `key: value` format, where:
    1. `key` is a path-selector string, without enclosing qoutes (`"`), to one of the metadoc/ruledoc supported fields.
    9. `value` is rego value of the type of the field as defined in the spec.

Example:

```rego

@id: "XYZ"
@title: "Pod has privileged container"
@description: "This policy checks for containers in the Pod that have the `securityContext.privileged` field set to `true`. Privileged containers can allow almost completely unrestricted host access."
@custom.severity: "High"
@entrypoints: ["deny"]
package app.xyz

@description: "containers returns all the containers in a Kubernetes resource (Pod, Deployment, etc.), if any. This includes init containers and regular containers"
@input: "the input docuemnt should be a valid Kubernetes resouce, or a valid Kubernetes Admission Review resource, of Pod, or another resource that defines a Pod, such as Deployment, ReplicaSet, StatefulSet, etc."
@output: "set of container objects"
@examples: ["app.xyz.test_privileged_denied", "app.xyz.test_unprivileged_allowed"]
deny[ret] {
  ...
}
```

> This syntax can also be written as-is inside a Rego comment, which is a least-friction path for adoption


</details>

**Comment by itaysk** [comment]
In today's meeting Torin pointed out a flaw in the design, where rule-level documentation is maintained in an object where the key is (path to) the rule name, but if the rule is partial, it's "parts" cannot be referred to unambiguously.

One suggestion to fix that was to mandate that ruledocs must be declared before the function that they describe. This way the order in which they are parsed will match the order of partial rules internally.

Another suggestion was to move the ruledoc to the rule level, so it's a rule-local variable with a known name. This solves the issue of associating rules with their docs but makes it harder to reference ruledoc fields from other rules. For example if there's an "entrypoint" rule that aggregates the results of "internal" rules. The workaround here can be to define the values at the policy (global) level, and refer to them from the internal rule-level variable. This (rule-level var) seems like the preferred option.

**Comment by itaysk** [comment]
Worth mentioning that this proposal is interfacing with another issue to add json schema validation: https://github.com/open-policy-agent/opa/issues/1449

**Comment by itaysk** [comment]
To summarize, this is what we are looking at:

```rego
package app.xyz

__rego__metadoc__ := {
  "id": "XYZ",
  "title": "Policy title",
  "description": "Policy description",
  "custom": {
    "something": "some value",
    "arrays too": [
      {
        "one",
        "two"
      }  
    ]
  },
  "entrypoints" : [
      "app.xyz.deny"
  ]
}

deny[ret] {
  __rego__ruledoc__ := {
    "description": "rule description",
    "input": "textual description of the input",
    "input_schema": "https://raw.githubusercontent.com/instrumenta/kubernetes-json-schema/master/v1.18.1-standalone/pod.json",
    "output": "textual description of output",
    "examples": [
        "app.xyz.test_privileged_denied",
        "app.xyz.test_unprivileged_allowed"
    ]
}
...
}
```

**Comment by tsandall** [comment]
@itaysk I was thinking about this a bit more and realized that defining annotations in Rego (as opposed to comments or a new language construct) would create some interesting challenges. Specifically, if the annotations are just plain Rego then the type checker (or the caller) needs to evaluate the policy before type checking runs. This creates a bit of a cyclic dependency because the evaluator assumes that policies have already been compiled. We could probably work around this but I could see it becoming complex.

What I think we should do is represent the annotations in comments w/ YAML for now: https://github.com/open-policy-agent/opa/pull/3123#discussion_r580294115

As part of that PR, we'll get the plumbing implemented to support this. We can extend the data model for annotations once we have the basics in place!

**Comment by itaysk** [comment]
We discussed restricting the rego rule to JSON (rather than dynamic rego) (It's actually still written in the proposal), I thought that would be best of both worlds: we could statically parse the json like you suggest, and rego rules could also refer to it once it's loaded. wdyt?

**Comment by tsandall** [comment]
I'm not sure the cost of implementing metadata-in-rego is going to be worth it. It would allow re-use in rules like you said but it only works in one direction. You couldn't re-use values inside the metadata itself, so you're still going to end up with duplication. E.g., suppose you wanted an identifier/code in the metadata and you needed to specify that identifier/code in more than one metadata document...you'd have to redefine it.

I don't disagree that the metadata-in-rego option is appealing. I'm just concerned about the implementation. E.g., it would be nice if the metadata introduced zero overhead (or at least, very little or optional overhead). With metadata-in-comments, it's relatively easy to ensure this: the parser can be configured to skip comment processing. Done! With metadata-in-rego, it's more complicated because now those values may even be needed at evaluation time.

I think it's just a tradeoff and since the re-use case is a bit speculative, I'd rather err on the side of a simple implementation. My guess is that most users won't care which one we pick and most users won't appreciate the beauty of the metadata-in-rego option (...but they will certainly complain about unintended side affects like performance degradation 😅 )

**Comment by itaysk** [comment]
ok! so should I'll catch up with the other PR and then post an updated regodoc example here

**Comment by itaysk** [comment]
@tsandall I'm recently witnessed 4 different projects that had similar request, that policies will have metadata for the purpose of 1) building external indexes and 2) incorporating in rule results.
concrete example: there are many policies, each policy has a name, category and severity. when the set of policies are evaluated, the violation result should includes the name, category and severity of the violating policy. in addition, the metadata will be use to auto-generate a nice web page for each policy.

The yaml in comment design will force these users to duplicate the same metadata, one time in yaml-comment, second time in a rego variable that will be used to construct the result object. This is not news for us, but I thought I'd drop this feedback here because I think this is a very common request, and we might be missing something if we don't address it fully.

> You couldn't re-use values inside the metadata itself

Yes this is true, but I think it's nice to have. Looking at these (real) project, none of them attempted to do so, In fact, one project even opted to keep the metadata in a data file (pure json). 

You folks obviously have seen more examples in the wild so you'll have a better perspective, but I thought it was my duty to give this feedback which might contribute some more data points to your view. I do think it's a very common requirement to include the policy metadata in result.

**Comment by petemounce** [comment]
I wish I'd put that together at the time I raised the ticket. I work in an engineering effectiveness team, and educate around automated testing and compliance. So, offering as another data point, the following.

We've had more success affecting customer behaviour (that is, quicker turnaround on people debugging then fixing failing tests for example) where the error/non-compliance messages are more actionable, because the most likely point of contact with compliance is a failure within CI.

We've found that it's more likely that a fixer will have a good experience (and go on to read wider documentation) if that failure is gentle and actionable. It's less likely a person will read documentation in isolation before making a change to achieve compliance first time.

I speculate that this effect might be more pronounced the more general the compliance rules are; if they are authored in-house, a fixer can ping internally. Whereas, if someone has included a set of general rules from an external author, it seems more likely an unfamiliar fixer will be more easily thwarted (no-one knowledgeable internally, perhaps also the person who included them isn't deeply familiar but did the include in a general "that seems like a good idea, go for it" sort of approach).

**Comment by tsandall** [comment]
@itaysk thanks for the extra info!

I think we can solve that re-use case by introducing a special variable that lets rules refer to the metadata attached to them (or the package). This would be similar to how languages like Python let you refer to metadata on functions from within the function itself (because everything is an object). For example:

```rego
# title: Some rule title
# severity: High
deny[msg] {
   # <some conditions>...
   msg := sprintf("%v: blah blah blah, [__meta__.severity])
}
```

In this case `__meta__` is a special variable that refers to the metadata defined on the rule. Something similar would work for package level metadata as well.

Once the initial implementation for metadata lands, we can do additional work to plumb the metadata into the evaluator.

**Comment by itaysk** [comment]
@tsandall that would work as well! thanks for listening ❤️

**Comment by tsandall** [comment]
I'm close to this issue in favor of #4113 which I just filed.
